### DISCUSSION TITLE
Sequence of model addition

### DISCUSSION THREAD

**Discussion by laat0003** [pastas/pastas#233] [Q&A] [answered]
Hi,

I am working in a semi-arid region in Australia with bore hydrographs that have two distinctive step increases to mean water table related to large flow events in an ephemeral stream. There is also an apparent increasing linear trend to the water table associated with consecutive smaller seasonal flow events. Unfortunately there are no data sets related to the stream itself only the monitoring bores. 

I set up a Pastas model with two step models (one for each of the big flow events), a linear trend model for the period between those events (approx. 11 years) in addition to a stress model with precipitation and potential evaporation. Long story short, the sequence of model additions has a noticeable impact on the final solution. If I don't add the linear trend first then the first step model overshoots so that and I end up with a linear trend of 0.0 and a poor fit. Adding the linear trend first solves the problem. I am not sure if this is an issue but I am curious if there is a recommended model addition sequence when using multiple models? I am happy to provide the data if needed. It could also be used to demonstrate implementing the step and linear models. Chances are you will do a better job than me with it anyway :)

Awesome work by the way.

**Comment by dbrakenhoff** [comment]
Hi @laat0003, 

Thanks for submitting your issue. It sounds interesting and is definitely worth looking into. Ideally, the order of added stressmodels shouldn't influence the solution (I think?). 

A working example would be quite helpful to start our investigation! So if you're okay with uploading zipfile that contains the data and a small script that shows the difference between the two models that would be great!

**Comment by laat0003** [comment]
[Pastas_test.zip](https://github.com/pastas/pastas/files/5068914/Pastas_test.zip)

After some more testing I noticed the error is somewhat inconsistent but I managed to reproduce it a few times. In the uploaded zipfile there is an html version of the notebook showing the disparity in results. I've included the files I am using to work with and a copy of the notebook itself.

Thanks for the quick response.

**Comment by dbrakenhoff** [comment, marked as answer]
Thanks for the files!

I ran your notebook on my PC (with the same pastas version v0.14.0) and I could not reproduce your results entirely. I'm getting a similar fit every time, where the linear trend doesn't seem to do much. I do see clear differences between the two models though.

When the linear trend is added first I get a slope of 0. When it is added later, I get a very large number (order 10^5). But this doesn't do much in either case because `tstart` and `tend` are set to some strange values in the optimization. In the first case `tend` lies before `tstart`. In the second case, both lie far into the future...

So something isn't working well when optimizing the tstart and tend. I can sort of imagine that optimizing these dates is a challenge for leastsquares. If I turn off the optimization of those parameters I get much better results. This can be done with the following code prior to adding the Linear stressmodel to the model:
```python
sm = ps.LinearTrend('02-2006','02-2017')
sm.parameters.loc["linear_trend_tstart", "vary"] = False
sm.parameters.loc["linear_trend_tend", "vary"] = False
```

I'll have to check with the other developers if optimizing these dates is at all feasible... Maybe they know more about this. I think it might also make sense to set those dates to not be optimized by default.

In the meantime I would recommend not optimizing those parameters. Hopefully that helps for now!

**Comment by laat0003** [comment]
Great! Thanks very much for that. Yes, the linear trend model is working much better now. Very much appreciate your time looking into this. 

As an FYI, the following error is flagged, albeit inconsistently when the linear trend model doesn't work as expected. 

C:\ProgramData\Anaconda3\lib\site-packages\pastas\solver.py:264: RuntimeWarning: invalid value encountered in double_scalars np.sqrt(pcov.loc[i, i] * pcov.loc[j, j])

**Comment by raoulcollenteur** [comment]
Thanks for the interesting example! 

The error you mentioned above (in solver.py) usually occurs when the standard errors cannot be determined. You may try LmfitSolve, or fix some parameters that are hard to estimate.

If you have an example notebook that you can/want to share on the Github and [Docs ](https://pastas.readthedocs.io/en/latest/examples/index.html) this would be great! I think the linear trend and step models are still missing from the documentation.

Cheers,
Raoul
